### PR TITLE
Add ability to focus actions and view their state snapshots.

### PR DESCRIPTION
**Pull request by prayash**
#12 - This PR adds the ability to focus on an action in the actionList component. The user can click on any action to examine its state snapshot. Clicking on an already focused action will snap back to the currently checked out action.

action-focusing

### CODE REVIEW

File: src/devtools/domains/focused.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  }
+
+  focus(_old, id) {
+    return id
+  }
+
+  register() {
+    return {
+      ['checkout']: this.focus,
+      ['detail']: this.focus
```
Comment (nhunzaker) [inline src/devtools/domains/focused.js]: This could just be standard keys if you wanted. No need to bracket em.

File: src/devtools/effects/events.js [hunk 2]
```diff
@@ -45,13 +46,18 @@ class Events {
...
   register() {
     return {
       ['toggle']: this.toggle,
       ['remove']: this.remove,
       ['checkout']: this.checkout,
       ['revert']: this.revert,
-      ['commit']: this.commit
+      ['commit']: this.commit,
+      ['detail']: this.detail
     }
```
Comment (nhunzaker) [inline src/devtools/effects/events.js:61]: Ah I see now. This is all probably my fault. Haha.

**Review by nhunzaker — APPROVED** [review]
So this just focuses the list, there's no additional stuff going on yet?

**Comment by prayash** [comment]
Yeah focusing on the list and updating the state snapshot pane on the right with the focused action's state snapshot. Will upload GIF soon.

File: src/backend/history.js [hunk 3]
```diff
@@ -28,6 +28,10 @@ export function trackHistory(hook, bridge) {
 
       history.checkout(action)
     })
+
+    bridge.on(`detail:${action.id}`, () => {
+      bridge.send('snapshot', JSON.stringify(repo.archive.get(action)))
+    })
```
Comment (efatsi) [inline src/backend/history.js]: should add `bridge.removeAllListeners('detail:${action.id}')` below as well

File: src/devtools/domains/history.js [hunk 4]
```diff
@@ -33,12 +36,24 @@ class History {
       }
     }
 
-    return { ...newHistory, list }
+    return { ...newHistory, list, focused }
+  }
+
+  focus(_old, id) {
```
Comment (efatsi) [inline src/devtools/domains/history.js]: an underscore is used to denote a variable that won't be used (both to future developers and often to linters/compilers), since you're using the previous state here, i'd just go with `old`

File: src/devtools/views/layout.js [hunk 5]
```diff
@@ -24,9 +24,11 @@ class Layout extends React.Component {
       return null
     }
 
+    let { history } = this.props
+
     return (
       <div className={css.rail}>
-        <ActionList history={this.props.history} />
+        <ActionList focused={history.focused} history={history} />
```
Comment (efatsi) [inline src/devtools/views/layout.js]: I think you can revert the changes here since you're pulling `focused` out of `history` in your `ActionList` anyway.

File: src/devtools/domains/history.js [hunk 6]
```diff
@@ -33,12 +36,24 @@ class History {
...
+    let focused = null
```
Comment (efatsi) [inline src/devtools/domains/history.js]: i believe when defining variables that will definitely be updated later, simply going with `let focused` is preferred

File: src/devtools/domains/history.js [hunk 7]
```diff
@@ -33,12 +36,24 @@ class History {
...
+    if (_old.focused === id) {
+      focused = _old.head
+    } else {
+      focused = id
+    }
```
Comment (efatsi) [inline src/devtools/domains/history.js]: since you have the ternary check `send('detail', isFocused ? head : action.id)`, is this switch even necessary? or can this just be something like:
```js
focus(old, id) {
  return { ...old, focused: id }
}
```

File: src/devtools/views/actionList/index.js [hunk 8]
```diff
@@ -1,13 +1,33 @@
...
+  return (
+    <Item
+      action={action}
+      isFocused={isFocused}
+      head={head}
+      key={action.id}
+      onClick={() => send('detail', isFocused ? head : action.id)}
+    />
+  )
+})
```
Comment (efatsi) [inline src/devtools/views/actionList/index.js]: @nhunzaker what are your thoughts on this? The `Application` presenter is pulling out everything from the model, and then various view components just pass down that data through the ranks. At what point do you start to rethink about the way these components live such that they become their own presenters that pull out the pieces of state only they care about?

File: src/devtools/domains/history.js (hunk 7 above)
Comment (prayash) [reply src/devtools/domains/history.js]: Yeah that was silly of me. I'll keep the ternary and throw out the if/else.

File: src/devtools/views/layout.js (hunk 5 above)
Comment (prayash) [reply src/devtools/views/layout.js]: You're right! Reverted that.

File: src/devtools/domains/history.js [hunk 9]
```diff
@@ -33,12 +36,18 @@ class History {
...
     }
 
-    return { ...newHistory, list }
+    return { ...newHistory, list, focused }
+  }
+
+  focus(old, id) {
+    let focused = id
+    return { ...old, focused }
   }
```
Comment (nhunzaker) [inline src/devtools/domains/history.js]: Style nitpick: what do you think of naming the second argument `focused` and nixing line 43? Like:

```
focus (old, focused) {
  return { ...old, focused }
}
```

File: src/devtools/views/actionRegion.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
+import React from 'react'
+import withSend from 'microcosm/addons/with-send'
+
+export default withSend(
+  class ActionRegion extends React.Component {
```
Comment (prayash) [inline src/devtools/views/actionRegion.js:5]: @efatsi We ended up abstracting that withSend out to a wrapper component called ActionRegion so we can reuse it wherever we want instead of having a specialized ItemWrapper component just to send that action.

Comment (efatsi) [reply src/devtools/views/actionRegion.js:5]: cool stuff!

**Comment by efatsi** [comment]
👍

**Comment by prayash** [comment]
Sorry about that. That should definitely not have gone through. My mistake for not reviewing carefully. Looks like Nate has cleaned that stuff out this morning.
